File: dotnet/Infrastructure/RabbitMQ.cs
using System.Diagnostics;
using System;
using System.Text;
using dotnet.Services;
using RabbitMQ.Client;
using RabbitMQ.Client.Events;

namespace dotnet.Infrastructure
{
    public class RabbitMQ : IDisposable
    {
        private readonly ConnectionFactory _factoryConnection;
        private IConnection? _connection;
        private IChannel? _channel;
        private string _queueName;

        public RabbitMQ(IConfiguration configuration, PersonService personService)  
        {
            _factoryConnection = new ConnectionFactory
            {
                HostName = configuration["RabbitMQ:Host"]!,
                UserName = configuration["RabbitMQ:Username"]!,
                Password = configuration["RabbitMQ:Password"]!,
            };

            _queueName = configuration["RabbitMQ:QueueName"]!;
        }

        public async Task Init()
        {
            int retries = 5;
            while (retries > 0)
            {
                try
                {
                    _connection = await _factoryConnection.CreateConnectionAsync();
                    _channel = await _connection.CreateChannelAsync();
                    await _channel.QueueDeclareAsync(queue: _queueName, durable: true, exclusive: false, autoDelete: false, arguments: null);
                    Console.WriteLine($"🎯 Queue '${_queueName}' created with success.");

                    return;
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error -> {ex.Message}");
                    Console.WriteLine($"❌ Error to connect on RabbitMQ, trying again... (${ retries} remaining attempts)");
                    retries--;

                    Thread.Sleep(5000); // Wait 5s before trying again
                }
            }

            Console.WriteLine("❌ Unable to connect to RabbitMQ.");
            Environment.Exit(0);
        }

        public async Task ConsumeAsync(Func<string, Task> processMessageAsync, CancellationToken stoppingToken)
        {
            var consumer = new AsyncEventingBasicConsumer(_channel!);
            consumer.ReceivedAsync += async (model, ea) =>
            {
                var body = ea.Body.ToArray();
                var message = Encoding.UTF8.GetString(body);

                try
                {
                    await processMessageAsync(message);
                    await _channel!.BasicAckAsync(ea.DeliveryTag, false);
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"❌ Error to process message: {ex.Message}");
                    await _channel!.BasicNackAsync(ea.DeliveryTag, false, false);
                }
            };

            await _channel!.BasicConsumeAsync(queue: _queueName, autoAck: false, consumer: consumer);

            await Task.Delay(Timeout.Infinite, stoppingToken); // Keep thread running
        }

        public void Dispose()
        {
            _channel?.CloseAsync();
            _connection?.CloseAsync();
        }
    }
}

File: node/src/infrastructure/rabbitMQ.ts
import amqp, { Connection, Channel, ConsumeMessage, Options } from "amqplib";
import dotenv from "dotenv";

dotenv.config();

export class RabbitMQ {
  private connection!: Connection;
  private channel!: Channel;
  private readonly queue: string;
  private readonly url: string;

  constructor() {
    this.url = process.env.RABBITMQ_URL as string;
    this.queue = process.env.INGESTION_QUEUE_NAME as string;
  }

  // Connected and initializes the channel and queue
  public async init(): Promise<void> {
    let retries = 5;
    while (retries) {
      try {
        this.connection = await amqp.connect(this.url);
        this.channel = await this.connection.createChannel();

        await this.channel.assertQueue(this.queue, { durable: true });
        console.log(`Queue '${this.queue}' created with success.`);
        
        return;
      } catch (error) {
        console.error(`Error to connect on RabbitMQ, trying again... (${retries} remaining attempts)`);
        retries--;
        
        await new Promise((res) => setTimeout(res, 5000)); // Wait 5s before trying again
      }
    }

    console.error("Unable to connect to RabbitMQ.");
    process.exit(1);
  }

  // Posts a message to the queue
  public async publish(message: string, queue: string, options?: Options.Publish): Promise<void> {
    if (!this.channel) {
      throw new Error("Channel not initialized. Call init() first.");
    }

    this.channel.sendToQueue(queue, Buffer.from(message), { persistent: true, ...options });
    console.log(`Posted message: ${message}`);
  }

  // Creates a consumer that processes messages with the callback function
  public async consume(callback: (message: ConsumeMessage) => Promise<void>): Promise<void> {
    if (!this.channel) {
      throw new Error("Channel not initialized. Call init() first.");
    }

    await this.channel.consume(this.queue, async (message) => {
      if (message) {
        try {
          await callback(message);
          this.channel.ack(message);
        } catch (err) {
          console.error("Error processing message:", err);
          
          this.channel.nack(message, false, false);
        }
      }
    });

    console.log("Waiting for messages...");
  }

  // Closes the connection and the channel
  public async close(): Promise<void> {
    await this.channel.close();
    await this.connection.close();
    console.log("Connection to RabbitMQ closed.");
  }
}